### DISCUSSION TITLE
Why is my "Getting started with USART" example only working on 20 and 16 MHz

### DISCUSSION THREAD

**Discussion by hmeijdam** [SpenceKonde/megaTinyCore#700] [Q&A] [answered]
I was looking at the [Microchip appnote TB3216 "Getting Started with USART"](https://ww1.microchip.com/downloads/en/Appnotes/TB3216-Getting-Started-with-USART-90003216B.pdf) and tried out the example on page 9

Below is the sketch I copied from the appnote
Before getting it to work, I had to make a small change and reconfigure the CPU clock prescaler from its default settings. After that it works correctly, but only if I compile with 20 or 16MHz clock.
If I compile with 10 or 8 MHz I get the double baudrate and my loop also runs twice as fast.
When I add a #define F_CPU 16000000L  at the top and compile at 8MHz it works with my defined baudrate again. 
When I add a #define F_CPU 20000000L  at the top and compile at 10MHz it works with my defined baudrate again. 

I would like to understand why I am getting a higher baudrate and my loop is running faster if I compile at a lower CPU clockspeed. I had assumed that the F_CPU definition would take care of this as suggested by the application note. I must be missing something.

```
#define USART0_BAUD_RATE(BAUD_RATE) ((float)(F_CPU * 64 / (16 * (float)BAUD_RATE)) + 0.5)
#include <util/delay.h>

int main(void) {
  // reconfigure CPU clock prescaler
  CCP = CCP_IOREG_gc;    // enable write to protected IO registers (timed sequence)
  CLKCTRL_MCLKCTRLB = 0; // turn off prescaler (and the default factor 6 division)

  USART0_init();
  while (1)
  {
    USART0_sendString("Hello World!\r\n");
    _delay_ms(1000);
  }
}

void USART0_init(void) {
  //  VPORTB_DIR |= PIN2_bm;   // set pin PB2 as output (use VPORT to save some flash instead of)
  //  PORTB.DIR |= PIN2_bm;    // or like in the application note
     pinMode(PIN_HWSERIAL0_TX, OUTPUT); // or if you want the core to figure out the pin (uses more memory)

  USART0.BAUD = (uint16_t)USART0_BAUD_RATE(9600);
  USART0.CTRLB |= USART_TXEN_bm;
}

void USART0_sendChar(char c) {
  while (!(USART0.STATUS & USART_DREIF_bm))
  { }
  USART0.TXDATAL = c;
}

void USART0_sendString(char *str) {
  for (size_t i = 0; i < strlen(str); i++)     USART0_sendChar(str[i]);
}
```

**Comment by SpenceKonde** [comment]
I suspect you' see a redefinition of F_CPU warning in the console output too..... 

The F_CPU selected from the tools menu is passed to the compiler as a -DF_CPU=20000000 for all .c/.cpp files.

**Comment by hmeijdam** [comment]
Yes, as I expected when I redefine the F_CPU I get this warning

```
 warning: "F_CPU" redefined
 #define F_CPU 20000000L

```
But without redefining it from 10 to 20 MHz I get the double baudrate and double loop speed.

**Comment by SpenceKonde** [comment]
Exactly - delay.h uses cycle c ounring to times the delays. So it delays
for half as many cycles to get what it thinks is 1s

The core is really not designed to be used the way you're doing it. Its
meant to work well when users use theb arduino api.

I looked into whether f_cpu could be changed ar runtime gracefully. Answer
is no, very no. It would destroy performance and require like 300 code mods
and the bloat would be unreal. Not going 5o happen!

You wa n t to set the clock speed to the speed it will be running at.

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Comment by hmeijdam** [comment]
I am not at all asking for any changes to be made to the core, And definitely not code size increase. Then I would have posted as issue. My question is about seeking to understand what is happening.
Why does the core have the timing right when I select 16 or 20 MHz in the menu while without making any code modifications the timing is off by a factor two when I select 8 or 10 MHz

You mention "you want to set the clock speed to the speed it will be running at" 
Isn't that what I do by selecting the speed in the menu?

Or should I stay away from re-configuring the clock prescaler?
I did that also, but the I have to redefine F_CPU into #define F_CPU 20000000L/6 to get the selected baudrate right

Or should I assume that when I do not use Setup{} and Loop{} the core does not set the CPU speed at all, and always keeps it at the highest value? While that's also confuses me a bit as F_CPU had a value, or I would not have gotten the redefine warning.

So just want to know what is actually going on, when I use above Microchip example code.

**Comment by SpenceKonde** [comment, marked as answer]
If you set the clock speed to 8 or 10, but then you change the prescaler, it is no longer running at the speed it was compiled for. All assumptions about timing that the core makes are violated. 
The core does not adjust the clock speed if you override main - you have to do that if you don;t want the /6 prescaler (which you definitelty don't, because the core does not support it and there is no option for the clock speed that will get correct timing). 

But you must always tell it what speed the code will be running at accurately, otherwise it will run at the wrong speed. The huge overhead I described is whatr would be required for the code to figure out and compensate for the speed not being known accurately at compile time. Tons of things that currenjtly optiomize awaty to constants could not do so - it would be a nightmare. If you will be running at 20 MHz, you MUST tell the core you are running at that speed, not 10. 

Additionally, it is worth noting that the example as presented requires that F_CPU be defined somewhere - the chip is not smart enough to figure out what speed it is running at. This is somewhere that it's best that we have to tell it the answer. Because whose to say that when set to the 20 MHz osc with no prescaler, that 2-series chip isn;t actually running with the pedal to the metal at 32 MHz via the calibration register? (See: Ref_Tuning.md )

**Comment by SpenceKonde** [comment]
The core and all associated code takes you at your word when you tell it the clock speed!

**Comment by hmeijdam** [comment]
So if I understand you correctly, when I select in the menu to compile for 10MHz, but I do not use Setup{} and Loop{} it is only compiled for 10 MHz, with al its time calculations based on the "#ifdef F_CPU 10000000L", but it will actually still running at 20Mhz
I would have to manually get rid of the /6 prescaler and manually set the clock speed from 20MHz to 10 MHz

Like what I see in wiring.c
```

  /******************************** CLOCK STUFF *********************************/
#if defined(CLOCK_TUNE_INTERNAL)
  void tune_internal(void);
#endif

void __attribute__((weak)) init_clock() {
  #ifndef CLOCK_SOURCE
    #error "CLOCK_SOURCE not defined. CLOCK_SOURCE must be either 0 (internal) or 2 (external clock)"
  #endif
  #if (CLOCK_SOURCE == 0)
    #if (defined(CLOCK_TUNE_INTERNAL))
      tune_internal(); // Will be inlined as only called once. Just too long and ugly to put two implementations in middle of this.
    #else
      #if (F_CPU == 20000000)
        /* No division on clock */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, 0x00);
      #elif (F_CPU == 16000000)
        /* No division on clock */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, 0x00);
      #elif (F_CPU == 10000000) // 20MHz prescaled by 2
        /* Clock DIV2 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_2X_gc));
      #elif (F_CPU == 8000000) // 16MHz prescaled by 2
        /* Clock DIV2 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_2X_gc));
      #elif (F_CPU == 5000000) // 20MHz prescaled by 4
        /* Clock DIV4 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_4X_gc));
      #elif (F_CPU == 4000000) // 16MHz prescaled by 4
        /* Clock DIV4 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_4X_gc));
      #elif (F_CPU == 1000000) // 16MHz prescaled by 16
        /* Clock DIV16 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_16X_gc));
      #else
        #ifndef F_CPU
          #error "F_CPU not defined"
        #else
          #error "F_CPU defined as an unsupported value for untuned internal oscillator"
        #endif
      #endif
```

**Comment by hmeijdam** [comment]
Thanks for helping to solve my confusion. This works for all clock speeds in the menu.

```
#define USART0_BAUD_RATE(BAUD_RATE) ((float)(F_CPU * 64 / (16 * (float)BAUD_RATE)) + 0.5)
#include <util/delay.h>

int main(void) {
  // reconfigure CPU clock prescaler
//  CCP = CCP_IOREG_gc;    // enable write to protected IO registers (timed sequence)
//  CLKCTRL_MCLKCTRLB = 0; // turn off prescaler (and the default factor 6 division)

     #if (F_CPU == 20000000)
        /* No division on clock */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, 0x00);
      #elif (F_CPU == 16000000)
        /* No division on clock */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, 0x00);
      #elif (F_CPU == 10000000) // 20MHz prescaled by 2
        /* Clock DIV2 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_2X_gc));
      #elif (F_CPU == 8000000) // 16MHz prescaled by 2
        /* Clock DIV2 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_2X_gc));
      #elif (F_CPU == 5000000) // 20MHz prescaled by 4
        /* Clock DIV4 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_4X_gc));
      #elif (F_CPU == 4000000) // 16MHz prescaled by 4
        /* Clock DIV4 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_4X_gc));
      #elif (F_CPU == 1000000) // 16MHz prescaled by 16
        /* Clock DIV16 */
        _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_16X_gc));
      #else
        #ifndef F_CPU
          #error "F_CPU not defined"
        #else
          #error "F_CPU defined as an unsupported value for untuned internal oscillator"
        #endif
      #endif


  USART0_init();
  while (1)
  {
    USART0_sendString("Hello World!\r\n");
    _delay_ms(1000);
  }
}

void USART0_init(void) {
  //  VPORTB_DIR |= PIN2_bm;   // set pin PB2 as output (use VPORT to save some flash instead of)
  //  PORTB.DIR |= PIN2_bm;    // or like in the application note
     pinMode(PIN_HWSERIAL0_TX, OUTPUT); // or if you want the core to figure out the pin (uses more memory)

  USART0.BAUD = (uint16_t)USART0_BAUD_RATE(9600);
  USART0.CTRLB |= USART_TXEN_bm;
}

void USART0_sendChar(char c) {
  while (!(USART0.STATUS & USART_DREIF_bm))
  { }
  USART0.TXDATAL = c;
}

void USART0_sendString(char *str) {
  for (size_t i = 0; i < strlen(str); i++)     USART0_sendChar(str[i]);
}
```
I still need to get used to the fact that changing clockspeed is not fuse driven anymore, but can be changed on the fly.

**Comment by hmeijdam** [comment]
The reason why I override the main is to save some space if I want to print some debug data on small parts like the ATTiny204. The above sketch compiles to 187 bytes (9% of progmem) where below sketch compiles to 1634 bytes consuming 79% of progmem. 
```

void setup() {
Serial.begin(9600);
}

void loop() {
Serial.println("Hello World!");
delay(1000);
}
```

**Comment by SpenceKonde** [comment]
Correct, overriding main is saying "I'm doing everything, trust that I am
doing it right". If you told it you were running at 10 mhz, it it your
responsibility ti ensure that it's running at that speed if you override
Main.

 There are very few things thst get done for you. Dirty reset protection
does still happen (resets chip and clears reset cause (it gets copied to
GPIOR0) before main or class constructors if it sees no reset cause,
because that indicates the chip got back to the reset vector illegally, no
reset has occurred, and the peripherals could be configured arbitrarily,
hence no api functions can be guaranteed not to produce undefined behavior,
see ref_reset.md for very detailed description on a dirty reset happens and
why it's responsible for most arsuino hangs where you need to power cycle
or reset button to fix it, and how to override that behavior.). And before
that it sets up the stack pointer so you can call functions without
shitting yourself and clears R1 so you can use the == operator and really
basic things like that. Interrupts are not enabled.

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Comment by SpenceKonde** [comment]
Yeah - that would be mostly "print". 2.5.11 has shrunk the serial code
twice already.

I've toyed with a ThinSerial library which would have very limited
capabilities and not rely on print to save space, but not enough hours in
the day, ya know?

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by hmeijdam** [reply]
No worries I can save myself for now with the above approach. And maybe others can too now.
